### ISSUE TITLE
Cannot Restore Trainer

### ISSUE THREAD

**Issue by dimoibiehg** [Microsoft/CNTK#1353]
I got the following exception in calling `trainer.restore_from_checkpoint` in CNTK. 
> 'This' function is not equivalent (isomorphic) to the function restored from a checkpoint.

My restoring code comes in the below. These are the same as the structure of the creating the trainer and saving the `trainer.dnn` by `trainer.save_checkpoint("trainer.dnn")` as mentioned [in this document][1].
Also, I should mention that **all of my function and model are the same as what are stored trainer using save_checkpoint function** (in python).

 

    def evaluate(reader, model):
        criterion = create_criterion_function(model)
        criterion.replace_placeholders({criterion.placeholders[0]: Input(input_dim),
                                        criterion.placeholders[1]: Input(label_dim)})

        # training config
        epoch_size = 34  
        minibatch_size = 17
    
        # LR schedule over epochs 
        lr_per_sample = [0.003]*4+[0.0015]*24+[0.0003]
        lr_per_minibatch = [x * minibatch_size for x in lr_per_sample]
        lr_schedule = learning_rate_schedule(lr_per_minibatch, UnitType.minibatch, epoch_size)
    
        # Momentum
        momentum_as_time_constant = momentum_as_time_constant_schedule(70)
    
        learner = adam_sgd(criterion.parameters,
                           lr=lr_schedule, momentum=momentum_as_time_constant,
                           low_memory=True,
                           gradient_clipping_threshold_per_sample=15, gradient_clipping_with_truncation=True)

        trainer = Trainer(model, criterion.outputs[0], criterion.outputs[1], learner)
        trainer.restore_from_checkpoint("trainer.dnn")

    def do_test():
        reader = create_reader('Test.txt', is_training=False)
        model = create_model()
        evaluate(reader, model)

    do_test()

  [1]: https://github.com/Microsoft/CNTK/wiki/Load-pre-trained-checkpointed-model-and-continue-retraining

**Comment by sayanpa** [comment]
Not sure if posted the same question on stackoverflow that I answered earlier. Here is the link: http://stackoverflow.com/questions/41783613/this-function-is-not-equivalent-isomorphic-to-the-function-restored-from-a-c

**Comment by eldakms** [comment]
Currently, if you change your script between the time when the checkpoint is taken and the time of restore - you will probably see this error. It is very restrictive currently - even if you change the order of the declaration of a input/output or the order of node creation (even the network is actually the same).

**Comment by dimoibiehg** [comment]
@sayanpa that was me! I didnot get my answer yet and I think it is a real issue in CNTK if I cannot restore , my checkpoint when I need it, not exactly after the saving the checkpoint!

**Comment by dimoibiehg** [comment]
@eldakms I know this, but I want to save the checkpoint and restore it in another session of the code, I think it is a natural requirement. Indeed, if I cannot to restore my checkpoint in a particular time using the same functions, it means I can't use CNTK in a real application! am I right?

**Comment by eldakms** [comment]
if you do not change .py file (how the network is created) between save/restore, you should be able to restore. If this does not work, please share the script that saves/restores the checkpoint.
Thanks! 
We are looking to remove this limitation, sorry for that.

**Comment by dimoibiehg** [comment]
Many Thanks. Indeed, it will not work, if even they are in the same .py file. Here is my code, and the error occurred in Line 113. 

    import math
    import numpy as np
    from cntk.blocks import default_options, LSTM, Placeholder, Input        # building blocks
    from cntk.layers import Embedding, Recurrence, Dense, BatchNormalization # layers
    from cntk.models import Sequential                                       # higher level things
    from cntk.utils import ProgressPrinter, log_number_of_parameters
    from cntk.io import MinibatchSource, CTFDeserializer
    from cntk.io import StreamDef, StreamDefs, INFINITELY_REPEAT, FULL_DATA_SWEEP
    from cntk import *
    from cntk.learner import adam_sgd, learning_rate_schedule
    
    # from cntk.device import set_default_device, gpu
    # set_default_device(gpu(0))
    
    # number of words in vocab, and intent labels
    vocab_size = 10; num_intents = 3
    
    # model dimensions
    input_dim  = 10
    label_dim  = 3
    emb_dim    = 150
    hidden_dim = 300
    
    def create_model():
        with default_options(initial_state=0.1):
            return Sequential([
                Embedding(emb_dim),
                Recurrence(LSTM(hidden_dim), go_backwards=False),
                Dense(num_intents)
            ])
    
    # # peek
    # model = create_model()
    # print(len(model.layers))
    # print(model.layers[0].E.shape)
    # print(model.layers[2].b.value)	
    
    
    def create_reader(path, is_training):
        return MinibatchSource(CTFDeserializer(path, StreamDefs(
             plays   = StreamDef(field='plays', shape = input_dim,  is_sparse=True), 
             choices = StreamDef(field='choices', shape = label_dim, is_sparse=False)
         )), randomize=is_training, epoch_size = INFINITELY_REPEAT if is_training else FULL_DATA_SWEEP)
     
    # # peek
    # reader = create_reader('Plays.txt', is_training=True)
    # print(reader.streams.keys())
    
    def create_criterion_function(model):
        labels = Placeholder(name='labels')
        ce   = cross_entropy_with_softmax(model, labels)
        errs = classification_error      (model, labels)	
        return combine ([ce, errs]) # (features, labels) -> (loss, metric)
    
    
    model = create_model()
    criterion = create_criterion_function(model)
    criterion.replace_placeholders({criterion.placeholders[0]: Input(input_dim),
                                    criterion.placeholders[1]: Input(label_dim)})
    
    # training config
    epoch_size = 34  
    minibatch_size = 17
        
    # LR schedule over epochs 
    lr_per_sample = [0.003]*4+[0.0015]*24+[0.0003]
    lr_per_minibatch = [x * minibatch_size for x in lr_per_sample]
    lr_schedule = learning_rate_schedule(lr_per_minibatch, UnitType.minibatch, epoch_size)
        
    # Momentum
    momentum_as_time_constant = momentum_as_time_constant_schedule(70)
        
    learner = adam_sgd(criterion.parameters,
                       lr=lr_schedule, momentum=momentum_as_time_constant,
                       low_memory=True,
                       gradient_clipping_threshold_per_sample=15, gradient_clipping_with_truncation=True)
    
    
    def train(reader, model, max_epochs=16):
    
        # trainer
        trainer = Trainer(model, criterion.outputs[0], criterion.outputs[1], learner)
    
        # process minibatches and perform model training
        log_number_of_parameters(model)
        progress_printer = ProgressPrinter(tag='Training')
        #progress_printer = ProgressPrinter(freq=100, first=10, tag='Training') # more detailed logging
    
        t = 0
        for epoch in range(max_epochs):         # loop over epochs
            epoch_end = (epoch+1) * epoch_size
            while t < epoch_end:                # loop over minibatches on the epoch
                data = reader.next_minibatch(minibatch_size, input_map={  # fetch minibatch
                    criterion.arguments[0]: reader.streams.plays,
                    criterion.arguments[1]: reader.streams.choices
                })
                #print(data)
                trainer.train_minibatch(data)                                     # update model with it
                t += data[criterion.arguments[1]].num_samples                     # samples so far
                progress_printer.update_with_trainer(trainer, with_metric=True)   # log progress
            loss, metric, actual_samples = progress_printer.epoch_summary(with_metric=True)
        trainer.save_checkpoint("trainer.dnn")
        #trainer.restore_from_checkpoint("trainer.dnn")
    
    def do_train():
        reader = create_reader('Plays.txt', is_training=True)
        train(reader, model, 5)
    
    def evaluate(reader):
    
        # trainer
        trainer = Trainer(model, criterion.outputs[0], criterion.outputs[1], learner)
        trainer.restore_from_checkpoint("trainer.dnn")
    	
        evaluator = trainer 
        progress_printer = ProgressPrinter(tag='Evaluation')
    
        while True:
            minibatch_size = 17
            data = reader.next_minibatch(minibatch_size, input_map={  # fetch minibatch
                criterion.arguments[0]: reader.streams.plays,
    			criterion.arguments[1]: reader.streams.choices
            })
            print(data)
            if not data:                                 # until we hit the end
                break
            metric = evaluator.test_minibatch(data)
            progress_printer.update(0, data[criterion.arguments[1]].num_samples, metric) # log progress
        loss, metric, actual_samples = progress_printer.epoch_summary(with_metric=True)
    
        return loss, metric
    
    def do_test():
        reader = create_reader('Test.txt', is_training=False)
        #model = create_model()
        do_train()
        evaluate(reader)
    do_test()

**Comment by eldakms** [comment]
Could you please try this on the latest master/v2 binaries. If functions are isomorphic the restore should work. Closing for now due to long inactivity, feel free to reopen if the issue is still present.
